### PR TITLE
Display Internet Connection Error inside Clone Repository Dialog

### LINKED ISSUE #2796: surface network error when fetching repositories in clone dialog
We should surface errors in the clone repo modal if there is no internet connection. The loading icon attempts to load repos and then eventually stops when it realizes there is no connection, however the user is not notified.

screen shot 2017-09-20 at 13 37 10

This applies to the Github and Enterprise tabs, as the URL shows the following error, which isn't too user-friendly either:

screen shot 2017-09-20 at 13 37 46

tested v1.0.0, mac

### PR DESCRIPTION
**Pull request by Daniel-McCarthy**
## Overview

**Closes #2796**

~❗️ **#6278 should be merged before this one gets merged because there will likely be some conflicting things** ❗️~

## Description

- Renders an internet connection error when repositories fail to load in the `Clone Repository` dialog.
Additionally, a "Try again?" button has been added to allow the user to initiate an attempt to reload the repositories.

Before:
image

After:
image
___
Regarding the implementation of these changes, one part I would consider changing but did not initially:
When the `Clone Repository` dialog attempts to load the repositories, it calls [`api.fetchRepositories()`](https://github.com/desktop/desktop/blob/de96ea63e8c4410ef6e2dca82e9163e3bbec2321/app/src/lib/api.ts#L273-L283). This function returns null when it encounters an error, and logs the error. So the caller can detect when loading the repositories failed due to an error, but not which error caused it. Therefore the current implementation can potentially display this "No internet connection" error when it fails due to a different issue.

If this api function instead returned an error code instead of null we could then make it only display the internet connection error when it is confirmed to be an internet issue. However, there may be another way to check this that I am not aware of.

## Release notes

Notes: An error will now display in the Clone Repository dialog when internet connection issues occur.

### CODE REVIEW

**Comment by billygriffin** [comment]
@Daniel-McCarthy 🎉 ❤️  Can I go ahead and add `ready-for-review` here?

**Comment by Daniel-McCarthy** [comment]
@billygriffin Yes, that would be great! It's ready for the first round of review.

**Comment by tierninho** [comment]
Tested the UI changes, and LGTM!

**Comment by iAmWillShepherd** [comment]
@Daniel-McCarthy [these](https://developer.github.com/v3/#client-errors) are the 3 types of errors we'd get back from the API. We could open an issue to create a new `IAPIError` interface and update our API functions to return either a `Promise` or the new type. Doing this will give us a bit more control of how we handle exceptional behavior. 

Let me know if you think that would be beneficial for this PR.

cc @desktop/engineering for thoughts on whether we should return an `APIResult` (made up name since each function returns a different type) or `APIError` instead of `APIResult` or `null`.

File: app/src/ui/clone-repository/clone-github-repository.tsx [hunk 1]
```diff
@@ -171,6 +192,11 @@ export class CloneGithubRepository extends React.Component<
     )
   }
 
+  private reloadAndRenderRepositories = () => {
+    this.loadRepositories(this.props.account)
+    this.renderRepositoryList()
```
[the remark this answers is not in the document]
Comment (Daniel-McCarthy) [reply app/src/ui/clone-repository/clone-github-repository.tsx]: Ah, great! Thank you for letting me know, I will commit the change now.

**Comment by billygriffin** [comment]
cc: @niik on this one as well as it touches the cloning dialog.

**Comment by Daniel-McCarthy** [comment]
> unfortunately this PR conflicts somewhat with #6278 so we're going to get that one landed first, and then pull this in afterward.

Sounds great. I'm more than happy to handle the merge conflicts on my end as soon as that PR gets merged.
___
> [these](https://developer.github.com/v3/#client-errors) are the 3 types of errors we'd get back from the API. We could open an issue to create a new `IAPIError` interface and update our API functions to return either a `Promise` or the new type. Doing this will give us a bit more control of how we handle exceptional behavior.
> 
> Let me know if you think that would be beneficial for this PR.

I'm not sure it would be vital for this pull request. I figured if there are other potential causes that would be useful to differentiate from for the user then it might be worth considering here. 

Depending on the information we can pass down we would potentially be able display a different error if the user's internet connection doesn't seem to be the problem. But I am not sure if the information we would get from the 3 client errors cover potential issues the user would want to know about.

Even if the issue is not their connection, the "Try Again?" button this new display provides is still very useful to the user as they will want to attempt to refresh regardless of the cause.

And thank you for sharing that information, it was very useful to read through.

**Comment by Daniel-McCarthy** [comment]
@billygriffin @niik 
Previously we had agreed to wait until #6278 was merged, then handle the merge conflicts here.
However, it appears that the conflicts are greater than expected. This pull request has some conflicting functionality with #6278.

This pull request uses the result of a `api.fetchRepositories` call to determine whether or not to display the internet connection warning. The changes made in #6278 use the same result to display that it is loading.

I'm not sure if there's a way we can decide which to show depending on the scenario, or if we would have to keep it how it is or substitute the loading display.

**Comment by billygriffin** [comment]
@Daniel-McCarthy Unfortunately I can't speak to the underlying technical tradeoff on this, so I'll have to defer to @niik or someone else in @desktop/engineering to weigh in. Thanks for succinctly explaining where the two are intersecting.

**Comment by Daniel-McCarthy** [comment]
I have looked into this further and was able to find a way to potentially reconcile the issue with the two displays that would potentially be shown if no repositories were loaded.

Currently if one goes to the Clone Repository dialog in the GitHub section and no repositories are loaded for the list it will show a display saying that no repositories could be found for the account and offering a link to attempt to refresh. (This has been added since the creation of this pull request.)

Shown here:
image

However, it does not imply any issues with internet connection, rather that the account itself has been checked for repositories.

I mentioned prior that there were potential issues regarding detecting whether or not the user has an internet connection. I have updated this by checking the `navigator.onLine` property. For Chrome it will return false if the user is not connected to a network. However, it is not a perfect indicator.

Essentially with the current updated implementation it will always show the above display unless they are not connected to another network. If they are not connected to a network it will display the original display design warning of no internet connection and offering a link to attempt to refresh, as seen here:

image

`navigator.onLine` may be sufficient for deciding which display to show. If there is a more reliable way to check for connection that would be preferred we could try that.

Additionally we can decide if this display is still wanted, given that there have been noticeable changes since the creation of the original issue.

**Comment by outofambit** [comment]
@Daniel-McCarthy I'm really liking this. Thanks for sticking with it! What do you think about using this text for the new error state introduced in this PR? (Where the second part refreshes the list.)

> No repositories found. Check again?

**Comment by iAmWillShepherd** [comment]
We think that using the `navigator.onLine` property is a good start to this solution, but we are concerned about its limitation to tell us whether there is internet access (see docs [here]( https://developer.mozilla.org/en-US/docs/Web/API/NavigatorOnLine/onLine)). At the moment, this property only tells us if the user is connected to a (W)LAN, **not** if they have internet connection
e.g., if I connect to my router that has no connection to the internet, the app will think that I have an internet connection when I don't.

We think you can check for this error to know for sure if there is internet access
screen shot 2019-02-19 at 4 21 23 pm

**Comment by Daniel-McCarthy** [comment]
@iAmWillShepherd Okay, sounds good. I'd like to get a better idea of the proposed changes here.

- In order for the function `renderNoItems` in `cloneable-repository-filter-list` to detect this `ERR_INTERNET_DISCONNECTED` Error it needs that information passed in as a prop alongside the `repositories` prop.

- To pass a boolean showing the error was thrown to `cloneable-repository-filter-list` as a prop, `blank-slate.tsx` and `clone-github-repository.tsx` would need to pass the information in from the `IAccountRepositories` account state.

- Then `api-repositories-store.tsx` needs to store the information after a calling `api.fetchRepositories()` where it gets the `repositories` prop that gets passed all the way down.

- In `api.ts` the `fetchRepositories` function does `log.warn()` with the error and returns null.
In order to make this work via the above changes, this function would have to return an Error type instead of null.
___
I'd like to be sure this is the set of changes we would need, or if there's a different way to access this information from the `renderNoItems` function that you had in mind?

Thanks for your time!

**Comment by iAmWillShepherd** [comment]
Just tagging https://github.com/desktop/desktop/issues/1128 as related in case we decide to tackle the general case of being offline

**Comment by iAmWillShepherd** [comment]
I spotted an interactions that could use some refinement. 

Black - Development version
White - Production version



When connection is lost and re-gained, the component does not know this, forcing me to have to click the refresh button (You get the message saying "Couldn't find any repositories..."). I think this can happen automagically as the current production version of the app does.

**Comment by billygriffin** [comment]
In talking this over with @Daniel-McCarthy, @nerdneha, and @niik, given the complexity here we all decided this probably went a little too far and our efforts (and @Daniel-McCarthy's) might better be prioritized on other things with more of a clear impact. I'm also going to close the associated issue because we haven't really heard this feedback from users as representing a real problem affecting more than a couple folks.

@Daniel-McCarthy Huge thanks for your efforts here and apologies we weren't as clear on the impact here as we could've been, and I'm excited to see what you work on next! 😄

**Pull request closed without merging**